File: Sources/SomnioUI/SpeechBubbleText.swift
import AppKit
import CoreGraphics
import Foundation

/// Pure-Foundation truncation + greedy word-wrap helpers for the speech bubble. The
/// font-driven measurement lives behind `defaultWidth(of:)` so unit tests can drive the
/// algorithm with a synthetic measurement closure.
public enum SpeechBubbleText {
    /// 150 px bubble width, matching the legacy `Sprechblase` template width.
    public static let bubbleWidth: CGFloat = 150

    /// Font size the wrap measurement and the renderer must share so pre-wrapped lines fit the
    /// width they were measured against.
    public static let fontSize: CGFloat = 10

    /// Greedy word-wraps `text` to fit `bubbleWidth` at System-10 metrics, then caps
    /// the result with `cap`. The wrap respects existing whitespace token boundaries
    /// only — long unbreakable words exceeding `bubbleWidth` are emitted as single
    /// lines (the renderer truncates them at draw time).
    @MainActor
    public static func wrap(
        _ text: String,
        maxLines: Int = 4,
        truncationGlyph: String = "..."
    ) -> [String] {
        wrap(text, maxLines: maxLines, truncationGlyph: truncationGlyph, widthOf: defaultWidth(of:))
    }

    /// Returns at most `maxLines` lines from `lines`. When the input is truncated, the
    /// last surviving line is appended with `truncationGlyph` to signal continuation.
    public static func cap(
        lines: [String],
        maxLines: Int = 4,
        truncationGlyph: String = "..."
    ) -> [String] {
        guard maxLines > 0 else { return [] }
        guard lines.count > maxLines else { return lines }
        var capped = Array(lines.prefix(maxLines))
        capped[capped.count - 1] += truncationGlyph
        return capped
    }

    /// Test seam — applies the greedy wrap algorithm using `widthOf` as the per-line
    /// width oracle, so unit tests can assert behavior without depending on AppKit
    /// font metrics.
    static func wrap(
        _ text: String,
        maxLines: Int,
        truncationGlyph: String,
        widthOf: (String) -> CGFloat
    ) -> [String] {
        let words = text.split(separator: " ", omittingEmptySubsequences: false).map(String.init)
        guard !words.isEmpty else { return [] }
        var lines: [String] = []
        var current = ""
        for word in words {
            let candidate = current.isEmpty ? word : current + " " + word
            if widthOf(candidate) <= bubbleWidth {
                current = candidate
            } else {
                if !current.isEmpty {
                    lines.append(current)
                }
                current = word
            }
        }
        if !current.isEmpty {
            lines.append(current)
        }
        return cap(lines: lines, maxLines: maxLines, truncationGlyph: truncationGlyph)
    }

    @MainActor
    private static func defaultWidth(of text: String) -> CGFloat {
        let font = NSFont.systemFont(ofSize: fontSize)
        return (text as NSString).size(withAttributes: [.font: font]).width
    }
}

File: Web/src/scene/speechBubbleText.ts
import { SOMNIO_CONSTANTS } from '@/core/constants'

/**
 * Mirror of `Sources/SomnioUI/SpeechBubbleText.swift`. The measurement is injected so the greedy
 * wrap is testable without font metrics, exactly as the Swift `widthOf` seam allows.
 */

export const BUBBLE_WIDTH = SOMNIO_CONSTANTS.speechBubbleWidthPixels
export const BUBBLE_FONT_SIZE = SOMNIO_CONSTANTS.speechBubbleFontSize

/** ASCII, matching the project-wide rule — never the Unicode ellipsis. */
export const TRUNCATION_GLYPH = '...'

/** Returns at most `maxLines` lines, marking truncation on the last surviving line. */
export function capLines(lines: string[], maxLines = 4, glyph = TRUNCATION_GLYPH): string[] {
  if (maxLines <= 0) return []
  if (lines.length <= maxLines) return lines
  const capped = lines.slice(0, maxLines)
  capped[capped.length - 1] += glyph
  return capped
}

/**
 * Greedy word wrap against `BUBBLE_WIDTH`.
 *
 * Respects existing whitespace boundaries only: a single unbreakable word wider than the bubble
 * is emitted as its own line and left for the renderer to truncate at draw time, rather than
 * being split mid-word.
 */
export function wrapSpeech(
  text: string,
  widthOf: (line: string) => number,
  maxLines = 4,
  glyph = TRUNCATION_GLYPH
): string[] {
  const words = text.split(' ')
  if (words.length === 0) return []
  const lines: string[] = []
  let current = ''
  for (const word of words) {
    const candidate = current === '' ? word : `${current} ${word}`
    if (widthOf(candidate) <= BUBBLE_WIDTH) {
      current = candidate
    } else {
      if (current !== '') lines.push(current)
      current = word
    }
  }
  if (current !== '') lines.push(current)
  return capLines(lines, maxLines, glyph)
}

/**
 * Canvas-backed measurement for the browser, at the same metrics the balloon is drawn with —
 * both sides must resolve to `SomnioConstants` or wrapped lines overflow the balloon body.
 */
export function canvasWidthMeasurer(): (line: string) => number {
  const context = document.createElement('canvas').getContext('2d')
  if (context === null) return (line) => line.length * BUBBLE_FONT_SIZE * 0.5
  context.font = `${BUBBLE_FONT_SIZE}px system-ui, sans-serif`
  return (line) => context.measureText(line).width
}

/** Lifetime rule from the legacy client: 2 s plus a second per line. */
export function bubbleLifetimeMs(lineCount: number): number {
  return 2000 + lineCount * 1000
}
